File: scripts/final-responsive-qa.mjs
import { chromium } from 'playwright';
import fs from 'node:fs/promises';
import path from 'node:path';
import { NAV_ITEMS } from '../src/data/roles.js';

const baseURL = process.env.QA_BASE_URL || 'http://127.0.0.1:4176/';
const outDir = path.resolve('qa-artifacts/final-responsive');
const storageKey = 'diagnosis-center-change-pack-v1-state';
const adminAuth = {
  role: 'admin',
  userName: 'System Admin',
  userId: 'AUTH-ADMIN',
  landing: 'admin-dashboard',
  linkedDoctorId: '',
  hospitalId: '',
  username: 'admin',
  loginAt: new Date().toISOString()
};

const viewports = [
  { name: 'phone', width: 390, height: 844, isMobile: true },
  { name: 'desktop', width: 1440, height: 900, isMobile: false }
];

const pages = [
  { id: 'login', label: 'Login', public: true, url: baseURL },
  ...NAV_ITEMS.map(item => ({ id: item.id, label: item.label, pageId: item.id, url: baseURL })),
  { id: 'public-report-verification', label: 'Public report verification', publicHash: '#/verify-report/SEC-ORD001-DEMO', url: `${baseURL}#/verify-report/SEC-ORD001-DEMO` },
  { id: 'public-patient-results', label: 'Public patient result portal', publicHash: '#/patient/results/PAT-001', url: `${baseURL}#/patient/results/PAT-001` }
];

function sanitize(name) {
  return name.replace(/[^a-z0-9_-]+/gi, '-').replace(/^-+|-+$/g, '').toLowerCase();
}

async function collectMetrics(page) {
  return page.evaluate(() => {
    const clientWidth = document.documentElement.clientWidth;
    const docScrollWidth = Math.max(document.documentElement.scrollWidth, document.body?.scrollWidth || 0);
    const main = document.getElementById('main-content');
    const mainScrollWidth = main ? main.scrollWidth : 0;
    const mainClientWidth = main ? main.clientWidth : 0;
    const visibleText = (document.body?.innerText || '').slice(0, 240);
    const offenders = [];
    for (const el of Array.from(document.body.querySelectorAll('*'))) {
      const style = window.getComputedStyle(el);
      if (style.display === 'none' || style.visibility === 'hidden' || style.opacity === '0') continue;
      const rect = el.getBoundingClientRect();
      if (rect.width === 0 || rect.height === 0) continue;
      const isFixedOffCanvas = style.position === 'fixed' && (rect.left >= clientWidth || rect.right <= 0);
      if (isFixedOffCanvas) continue;
      if (rect.right > clientWidth + 3 || rect.left < -3 || rect.width > clientWidth + 3) {
        offenders.push({
          tag: el.tagName.toLowerCase(),
          className: String(el.className || '').slice(0, 160),
          text: String(el.textContent || '').trim().replace(/\s+/g, ' ').slice(0, 100),
          rect: { left: Math.round(rect.left), right: Math.round(rect.right), width: Math.round(rect.width) }
        });
      }
      if (offenders.length >= 8) break;
    }
    const errorBoundaryVisible = /Something went wrong|Reset demo data and retry/i.test(document.body?.innerText || '');
    const loginVisible = /Sign in|Quick role login/i.test(document.body?.innerText || '');
    return {
      title: document.title,
      clientWidth,
      docScrollWidth,
      mainClientWidth,
      mainScrollWidth,
      bodyOverflow: docScrollWidth > clientWidth + 3,
      mainOverflow: main ? mainScrollWidth > mainClientWidth + 3 : false,
      errorBoundaryVisible,
      loginVisible,
      visibleText,
      offenders
    };
  });
}

async function run() {
  await fs.rm(outDir, { recursive: true, force: true });
  await fs.mkdir(outDir, { recursive: true });
  const browser = await chromium.launch({ headless: true, executablePath: '/usr/bin/chromium' });
  const results = [];
  for (const viewport of viewports) {
    const context = await browser.newContext({
      viewport: { width: viewport.width, height: viewport.height },
      isMobile: viewport.isMobile,
      deviceScaleFactor: viewport.isMobile ? 2 : 1,
      userAgent: viewport.isMobile
        ? 'Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1'
        : undefined
    });
    await context.route('**/api/**', route => route.fulfill({ status: 503, contentType: 'application/json', body: JSON.stringify({ ok: false, qa: 'offline-api-simulated' }) }));
    const consoleErrors = [];
    context.on('page', page => {
      page.on('console', msg => {
        if (['error'].includes(msg.type())) consoleErrors.push(msg.text());
      });
      page.on('pageerror', err => consoleErrors.push(err.message));
    });

    for (const appPage of pages) {
      const page = await context.newPage();
      const errorsBefore = consoleErrors.length;
      await page.addInitScript(({ storageKey, auth, pageId, publicPage }) => {
        window.localStorage.clear();
        if (!publicPage) {
          window.localStorage.setItem(storageKey, JSON.stringify({ auth, currentPage: pageId || 'admin-dashboard', ui: { sidebarOpen: false, toast: null } }));
        }
      }, { storageKey, auth: adminAuth, pageId: appPage.pageId, publicPage: appPage.public || appPage.publicHash });

      let status = 'passed';
      let notes = [];
      try {
        await page.goto(appPage.url, { waitUntil: 'networkidle', timeout: 20000 });
        await page.waitForTimeout(350);
        const screenshotPath = path.join(outDir, `${viewport.name}-${sanitize(appPage.id)}.png`);
        await page.screenshot({ path: screenshotPath, fullPage: true });
        const metrics = await collectMetrics(page);
        const newErrors = consoleErrors.slice(errorsBefore).filter(error => !/Failed to load resource/i.test(error));
        if (metrics.bodyOverflow) notes.push(`document horizontal overflow: ${metrics.docScrollWidth}px > ${metrics.clientWidth}px`);
        if (metrics.mainOverflow) notes.push(`main horizontal overflow: ${metrics.mainScrollWidth}px > ${metrics.mainClientWidth}px`);
        if (metrics.errorBoundaryVisible) notes.push('error boundary visible');
        if (!appPage.public && appPage.id !== 'login' && metrics.loginVisible) notes.push('unexpected login screen after auth seed');
        if (newErrors.length) notes.push(`console/page errors: ${newErrors.slice(0, 3).join(' | ')}`);
        if (notes.length) status = 'review';
        results.push({ viewport: viewport.name, page: appPage.id, label: appPage.label, status, notes, metrics, screenshot: screenshotPath });
      } catch (error) {
        results.push({ viewport: viewport.name, page: appPage.id, label: appPage.label, status: 'failed', notes: [error.message], metrics: null, screenshot: null });
      } finally {
        await page.close();
      }
    }

    // Interaction checks on dashboard for menus/drawers.
    const page = await context.newPage();
    await page.addInitScript(({ storageKey, auth }) => {
      window.localStorage.clear();
      window.localStorage.setItem(storageKey, JSON.stringify({ auth, currentPage: 'admin-dashboard', ui: { sidebarOpen: false, toast: null } }));
    }, { storageKey, auth: adminAuth });
    const interaction = { viewport: viewport.name, page: 'interaction-checks', label: 'Header / sidebar / notification / user menu', status: 'passed', notes: [] };
    try {
      await page.goto(baseURL, { waitUntil: 'networkidle', timeout: 20000 });
      await page.waitForTimeout(300);
      if (viewport.name === 'phone') {
        const menuButton = page.locator('button[aria-label*="Open navigation"], button:has-text("☰")').first();
        if (await menuButton.count()) {
          await menuButton.click();
          await page.waitForTimeout(150);
          const drawerVisible = await page.locator('[role="dialog"][aria-label*="navigation" i], aside').first().isVisible().catch(() => false);
          if (!drawerVisible) interaction.notes.push('mobile drawer did not become visible');
          await page.screenshot({ path: path.join(outDir, `${viewport.name}-interaction-sidebar-open.png`), fullPage: true });
          await page.keyboard.press('Escape').catch(() => {});
        } else {
          interaction.notes.push('mobile navigation button not found');
        }
      }
      const notificationButton = page.locator('button[aria-label*="notification" i]').first();
      if (await notificationButton.count()) {
        await notificationButton.click();
        await page.waitForTimeout(150);
        await page.screenshot({ path: path.join(outDir, `${viewport.name}-interaction-notifications.png`), fullPage: true });
        await page.keyboard.press('Escape').catch(() => {});
      } else {
        interaction.notes.push('notification button not found');
      }
      const userButton = page.locator('button[aria-label*="user" i], button:has-text("System Admin"), button:has-text("Admin")').first();
      if (await userButton.count()) {
        await userButton.click();
        await page.waitForTimeout(150);
        await page.screenshot({ path: path.join(outDir, `${viewport.name}-interaction-user-menu.png`), fullPage: true });
      } else {
        interaction.notes.push('user menu button not found');
      }
      const metrics = await collectMetrics(page);
      if (metrics.bodyOverflow) interaction.notes.push(`interaction overflow: ${metrics.docScrollWidth}px > ${metrics.clientWidth}px`);
      if (interaction.notes.length) interaction.status = 'review';
    } catch (error) {
      interaction.status = 'failed';
      interaction.notes.push(error.message);
    } finally {
      results.push(interaction);
      await page.close();
    }
    await context.close();
  }
  await browser.close();
  const summary = {
    generatedAt: new Date().toISOString(),
    baseURL,
    pageCount: pages.length,
    viewportCount: viewports.length,
    totals: {
      passed: results.filter(r => r.status === 'passed').length,
      review: results.filter(r => r.status === 'review').length,
      failed: results.filter(r => r.status === 'failed').length,
      total: results.length
    },
    results
  };
  await fs.writeFile(path.join(outDir, 'final-responsive-qa-results.json'), JSON.stringify(summary, null, 2));
  console.log(JSON.stringify(summary.totals, null, 2));
  const reviews = results.filter(r => r.status !== 'passed');
  if (reviews.length) {
    console.log('Items requiring review:');
    for (const item of reviews) console.log(`${item.viewport} ${item.page}: ${item.notes.join('; ')}`);
  }
}

run().catch(error => {
  console.error(error);
  process.exit(1);
});

File: scripts/final_responsive_qa.py
from playwright.sync_api import sync_playwright
import json, os, re, pathlib, time

BASE_URL = os.environ.get('QA_BASE_URL', 'http://127.0.0.1:4176/')
OUT_DIR = pathlib.Path('qa-artifacts/final-responsive')
STORAGE_KEY = 'diagnosis-center-change-pack-v1-state'
ADMIN_AUTH = {
    'role': 'admin', 'userName': 'System Admin', 'userId': 'AUTH-ADMIN', 'landing': 'admin-dashboard',
    'linkedDoctorId': '', 'hospitalId': '', 'username': 'admin', 'loginAt': time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime())
}
roles_text = pathlib.Path('src/data/roles.js').read_text()
items = re.findall(r"\{\s*id:\s*'([^']+)'\s*,\s*label:\s*'([^']+)'", roles_text)
# de-dupe keeping first; regex may match only nav items from NAV_ITEMS and possibly roles access? okay check.
seen=[]; nav=[]
for id_, label in items:
    if id_ not in seen and id_ not in ['doctor','receptionist','lab','scan','billing','admin']:
        seen.append(id_); nav.append((id_,label))

pages = [{'id':'login','label':'Login','public':True,'url':BASE_URL}]
pages += [{'id':id_, 'label':label, 'pageId':id_, 'url':BASE_URL} for id_,label in nav]
pages += [
    {'id':'public-report-verification','label':'Public report verification','publicHash':'#/verify-report/SEC-ORD001-DEMO','url':BASE_URL + '#/verify-report/SEC-ORD001-DEMO'},
    {'id':'public-patient-results','label':'Public patient result portal','publicHash':'#/patient/results/PAT-001','url':BASE_URL + '#/patient/results/PAT-001'},
]
viewports = [
    {'name':'phone','width':390,'height':844,'is_mobile':True},
    {'name':'desktop','width':1440,'height':900,'is_mobile':False},
]

def sanitize(name):
    return re.sub(r'[^a-z0-9_-]+','-',name, flags=re.I).strip('-').lower()

def seed_state_script(storage_key, auth, page_id=None, public=False):
    payload = None if public else {'auth': auth, 'currentPage': page_id or 'admin-dashboard', 'ui': {'sidebarOpen': False, 'toast': None}}
    return f"""
      window.localStorage.clear();
      if ({'true' if payload is not None else 'false'}) {{
        window.localStorage.setItem({json.dumps(storage_key)}, {json.dumps(json.dumps(payload))});
      }}
    """

def collect_metrics(page):
    return page.evaluate("""() => {
      const clientWidth = document.documentElement.clientWidth;
      const docScrollWidth = Math.max(document.documentElement.scrollWidth, document.body?.scrollWidth || 0);
      const main = document.getElementById('main-content');
      const mainScrollWidth = main ? main.scrollWidth : 0;
      const mainClientWidth = main ? main.clientWidth : 0;
      const visibleText = (document.body?.innerText || '').slice(0, 240);
      const offenders = [];
      for (const el of Array.from(document.body.querySelectorAll('*'))) {
        const style = window.getComputedStyle(el);
        if (style.display === 'none' || style.visibility === 'hidden' || style.opacity === '0') continue;
        const rect = el.getBoundingClientRect();
        if (rect.width === 0 || rect.height === 0) continue;
        const isFixedOffCanvas = style.position === 'fixed' && (rect.left >= clientWidth || rect.right <= 0);
        if (isFixedOffCanvas) continue;
        if (rect.right > clientWidth + 3 || rect.left < -3 || rect.width > clientWidth + 3) {
          offenders.push({
            tag: el.tagName.toLowerCase(),
            className: String(el.className || '').slice(0, 160),
            text: String(el.textContent || '').trim().replace(/\s+/g, ' ').slice(0, 100),
            rect: { left: Math.round(rect.left), right: Math.round(rect.right), width: Math.round(rect.width) }
          });
        }
        if (offenders.length >= 8) break;
      }
      const text = document.body?.innerText || '';
      const errorBoundaryVisible = /Something went wrong|Reset demo data and retry/i.test(text);
      const loginVisible = /Sign in|Quick role login/i.test(text);
      return {
        title: document.title,
        clientWidth,
        docScrollWidth,
        mainClientWidth,
        mainScrollWidth,
        bodyOverflow: docScrollWidth > clientWidth + 3,
        mainOverflow: main ? mainScrollWidth > mainClientWidth + 3 : false,
        errorBoundaryVisible,
        loginVisible,
        visibleText,
        offenders
      };
    }""")

OUT_DIR.mkdir(parents=True, exist_ok=True)
for p in OUT_DIR.glob('*'):
    if p.is_file(): p.unlink()

results=[]
with sync_playwright() as pw:
    browser = pw.chromium.launch(headless=True, executable_path='/usr/bin/chromium')
    for vp in viewports:
        context = browser.new_context(
            viewport={'width':vp['width'], 'height':vp['height']},
            is_mobile=vp['is_mobile'],
            device_scale_factor=2 if vp['is_mobile'] else 1,
            user_agent='Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1' if vp['is_mobile'] else None
        )
        context.route('**/api/**', lambda route: route.fulfill(status=503, content_type='application/json', body=json.dumps({'ok':False,'qa':'offline-api-simulated'})))
        console_errors=[]
        def setup_page(pg):
            pg.on('console', lambda msg: console_errors.append(msg.text) if msg.type == 'error' else None)
            pg.on('pageerror', lambda err: console_errors.append(str(err)))
        context.on('page', setup_page)
        for app_page in pages:
            page = context.new_page(); setup_page(page)
            err_before=len(console_errors)
            page.add_init_script(seed_state_script(STORAGE_KEY, ADMIN_AUTH, app_page.get('pageId'), app_page.get('public') or app_page.get('publicHash')))
            status='passed'; notes=[]; metrics=None; screenshot_path=None
            try:
                page.goto(app_page['url'], wait_until='networkidle', timeout=20000)
                page.wait_for_timeout(350)
                screenshot_path = str(OUT_DIR / f"{vp['name']}-{sanitize(app_page['id'])}.png")
                page.screenshot(path=screenshot_path, full_page=True)
                metrics = collect_metrics(page)
                new_errors=[e for e in console_errors[err_before:] if not re.search(r'Failed to load resource|404', e, re.I)]
                if metrics['bodyOverflow']:
                    notes.append(f"document horizontal overflow: {metrics['docScrollWidth']}px > {metrics['clientWidth']}px")
                if metrics['mainOverflow']:
                    notes.append(f"main horizontal overflow: {metrics['mainScrollWidth']}px > {metrics['mainClientWidth']}px")
                if metrics['errorBoundaryVisible']:
                    notes.append('error boundary visible')
                if not app_page.get('public') and app_page['id'] != 'login' and metrics['loginVisible']:
                    notes.append('unexpected login screen after auth seed')
                if new_errors:
                    notes.append('console/page errors: ' + ' | '.join(new_errors[:3]))
                if notes:
                    status='review'
            except Exception as e:
                status='failed'; notes.append(str(e))
            finally:
                results.append({'viewport':vp['name'], 'page':app_page['id'], 'label':app_page['label'], 'status':status, 'notes':notes, 'metrics':metrics, 'screenshot':screenshot_path})
                page.close()
        # interactions
        page = context.new_page(); setup_page(page)
        interaction={'viewport':vp['name'], 'page':'interaction-checks', 'label':'Header / sidebar / notification / user menu', 'status':'passed', 'notes':[]}
        page.add_init_script(seed_state_script(STORAGE_KEY, ADMIN_AUTH, 'admin-dashboard', False))
        try:
            page.goto(BASE_URL, wait_until='networkidle', timeout=20000); page.wait_for_timeout(300)
            if vp['name']=='phone':
                loc=page.locator('button[aria-label*="Open navigation"], button:has-text("☰")').first
                try:
                    if loc.count():
                        loc.click(); page.wait_for_timeout(150)
                        drawer_visible = page.locator('[role="dialog"][aria-label*="navigation" i], aside').first.is_visible()
                        if not drawer_visible: interaction['notes'].append('mobile drawer did not become visible')
                        page.screenshot(path=str(OUT_DIR/f"{vp['name']}-interaction-sidebar-open.png"), full_page=True)
                        page.keyboard.press('Escape')
                    else: interaction['notes'].append('mobile navigation button not found')
                except Exception as e: interaction['notes'].append('mobile drawer check error: '+str(e))
            try:
                nb=page.locator('button[aria-label*="notification" i]').first
                if nb.count():
                    nb.click(); page.wait_for_timeout(150); page.screenshot(path=str(OUT_DIR/f"{vp['name']}-interaction-notifications.png"), full_page=True); page.keyboard.press('Escape')
                else: interaction['notes'].append('notification button not found')
            except Exception as e: interaction['notes'].append('notification check error: '+str(e))
            try:
                ub=page.locator('button[aria-label*="user" i], button:has-text("System Admin"), button:has-text("Admin")').first
                if ub.count():
                    ub.click(); page.wait_for_timeout(150); page.screenshot(path=str(OUT_DIR/f"{vp['name']}-interaction-user-menu.png"), full_page=True)
                else: interaction['notes'].append('user menu button not found')
            except Exception as e: interaction['notes'].append('user menu check error: '+str(e))
            metrics=collect_metrics(page)
            if metrics['bodyOverflow']: interaction['notes'].append(f"interaction overflow: {metrics['docScrollWidth']}px > {metrics['clientWidth']}px")
            if interaction['notes']: interaction['status']='review'
        except Exception as e:
            interaction['status']='failed'; interaction['notes'].append(str(e))
        finally:
            results.append(interaction); page.close()
        context.close()
    browser.close()
summary={
    'generatedAt': time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime()),
    'baseURL': BASE_URL,
    'pageCount': len(pages),
    'viewportCount': len(viewports),
    'totals': {
        'passed': sum(1 for r in results if r['status']=='passed'),
        'review': sum(1 for r in results if r['status']=='review'),
        'failed': sum(1 for r in results if r['status']=='failed'),
        'total': len(results)
    },
    'results': results
}
(OUT_DIR/'final-responsive-qa-results.json').write_text(json.dumps(summary, indent=2))
print(json.dumps(summary['totals'], indent=2))
for r in results:
    if r['status']!='passed':
        print(f"{r['viewport']} {r['page']}: {'; '.join(r['notes'])}")
